src/oscar/apps/dashboard/catalogue/views.py

```diff
@@ -747,3 +759,238 @@ def get_context_data(self, *args, **kwargs):
     def get_success_url(self):
         messages.info(self.request, _("Product type deleted successfully"))
         return reverse("dashboard:catalogue-class-list")
+
+
+class AttributeOptionGroupCreateUpdateView(generic.UpdateView):
+
+    template_name = 'dashboard/catalogue/attribute_option_group_form.html'
+    model = AttributeOptionGroup
+    form_class = AttributeOptionGroupForm
+    attribute_option_formset = AttributeOptionFormSet
+
+    def process_all_forms(self, form):
+        """
+        This validates both the AttributeOptionGroup form and the
+        AttributeOptions formset at once making it possible to display all their
+        errors at once.
+        """
+        if self.creating and form.is_valid():
+            # the object will be needed by the attribute_option_formset
+            self.object = form.save(commit=False)
+
+        attribute_option_formset = self.attribute_option_formset(
+            self.request.POST, self.request.FILES, instance=self.object)
+
+        is_valid = form.is_valid() and attribute_option_formset.is_valid()
+
+        if is_valid:
+            return self.forms_valid(form, attribute_option_formset)
+        else:
+            return self.forms_invalid(form, attribute_option_formset)
+
+    def forms_valid(self, form, attribute_option_formset):
+        form.save()
+        attribute_option_formset.save()
+
+        return HttpResponseRedirect(self.get_success_url())
+
+    def forms_invalid(self, form, attribute_option_formset):
+        messages.error(self.request,
+                       _("Your submitted data was not valid - please "
+                         "correct the errors below"
+                         ))
+        ctx = self.get_context_data(form=form,
+                                    attribute_option_formset=attribute_option_formset)
+        return self.render_to_response(ctx)
+
+    form_valid = form_invalid = process_all_forms
+
+    def get_context_data(self, **kwargs):
+        ctx = super(AttributeOptionGroupCreateUpdateView, self).get_context_data(**kwargs)
+
+        if "attribute_option_formset" not in ctx:
+            ctx["attribute_option_formset"] = self.attribute_option_formset(
+                instance=self.object)
+
+        ctx["title"] = self.get_title()
+
+        if RelatedFieldWidgetWrapper.TO_FIELD_VAR in self.request.GET or RelatedFieldWidgetWrapper.TO_FIELD_VAR in self.request.POST:
```

Comment:
This logic - for handling the popups etc, is generic I assume. If so, can we move it into a resuable mixin that can, in future, be used for other related object views? Same for the other views below that are using the related object wrapper.